### ISSUE TITLE
BitfinexExchange claims there are no rate limits, but there are

### ISSUE THREAD

**Issue by cowwoc** [timmolter/XChange#2044]
BitfinexExchange's metadata returns `null` for both private and public rate limits, but if you fire a burst of requests you will get blocked by their server with an error message indicating you surpassed their limit.

https://bitcoin.stackexchange.com/a/36970/67917 and https://www.bitfinex.com/posts/188 touches upon this topic.

**Comment by timmolter** [comment]
If someone can figure out the rate limits by trial and error, it would be a valuable addition to the MetaData json object.

**Comment by cowwoc** [comment]
@timmolter Their documentation is quite clear on this point:

> If an IP address exceeds a certain number of requests per minute (between 10 and 90) to a specific REST API endpoint (ie. /ticker), the requesting IP address will be blocked for 10-60 seconds on that endpoint and the JSON response {"error": "ERR_RATE_LIMIT"} will be returned. Please note the exact logic and handling for such DDoS defenses may change over time to further improve reliability

Source: https://docs.bitfinex.com/v2/docs

V1 contains different wording but the same limit. Notice that now you can't predict the rate up-front (unless you want to limit yourself to 10 requests per minute). Instead you need to run a burst and back off in response to `ERR_RATE_LIMIT`. This is actually in line with how the HTTP protocol works (dynamic rates) so I expect others will do the same in the future.

I suggest throwing the same/consistent exception for this situation, across all exchanges.

**Comment by achavan51** [comment]
isn't there any other option to handle this from code side? instead of throwing the exception?

**Comment by timmolter** [comment]
The exchange API does not have this built in, and it is not part of the road map. The XChange API's boundary is where clients call endpoints on exchange servers. However the clients handle the responses is up to the client. In this case an **exception**al circumstance arose - the rate limit was exceeded, just one of many possible exceptions that can and will occur, and the client code needs to be able to handle all these circumstances.

**Comment by achavan51** [comment]
Yes, but in my case I need to handle this using the java code because every time throwing the exception will mess-up the things at my end and so I need the code example which can calculate the request time per minute and if the request limit will be >90 then it will just remove them and not hit this to the server.

**Comment by caleb-allen** [comment]
@achavan51 why not catch the RateLimitExceededException and handle it just as you said?
